test/integration/hotrestart_main.cc

```diff
@@ -0,0 +1,55 @@
+#include "common/stats/utility.h"
+
+#include "exe/main_common.h"
+
+#include "absl/debugging/symbolize.h"
+
+// NOLINT(namespace-envoy)
+
+/**
+ * Custom main() for hotrestart_test. This should be identical to
+ * source/exe/main.cc, except for the registration and increment of a new gauge
+ * specifically for hot_restart.test.sh.
+ *
+ * If this test starts failing to run or build, please ensure that all changes
+ * from source/exe/main.cc are incorporated in this file.
+ */
+int main(int argc, char** argv) {
+#ifndef __APPLE__
+  // absl::Symbolize mostly works without this, but this improves corner case
+  // handling, such as running in a chroot jail.
+  absl::InitializeSymbolizer(argv[0]);
+#endif
+  std::unique_ptr<Envoy::MainCommon> main_common;
+
+  // Initialize the server's main context under a try/catch loop and simply return EXIT_FAILURE
+  // as needed. Whatever code in the initialization path that fails is expected to log an error
+  // message so the user can diagnose.
+  try {
+    main_common = std::make_unique<Envoy::MainCommon>(argc, argv);
+
+    // This block is specific to hotrestart_main.cc, and should be the only
+    // functional difference with source/exe/main.cc.
+    Envoy::Server::Instance* server = main_common->server();
+    if (server != nullptr) {
+      // Creates a gauge that will be incremented once and then never touched. This is
+      // for testing parent-gauge accumulation in hot_restart_test.sh.
+      Envoy::Stats::Utility::gaugeFromElements(server->stats(),
+                                               {Envoy::Stats::DynamicName("hotrestart_test_gauge")},
+                                               Envoy::Stats::Gauge::ImportMode::Accumulate)
+          .inc();
+    }
```

Comment:
Instead of copying the entire body of main, wdyt about just having a utility helper that takes an optional "post server create" lambda that can be null in the real impl? That would avoid all of this copying and I think be pretty trivial to do?

Reply:
where would I register this post-server-create lambda? in a static ctor?

another thing I thought of, that wouldn't be too cumbersome and wouldn't affect prod execution flow, would be to move the contents of main.cc to main.h, and have in it:

```
  main_common = std::make_unique<Envoy::MainCommon>(argc, argv);
#ifdef POST_SERVER_CREATE_HOOK
  POST_SERVER_CREATE_HOOK(main_common->server())
#endif
```

Then main.cc would just have 
```
#include "source/exe/main.h"
```

WDYT?

Reply:
I just meant something like this:
```
int main_helper(int argc, char** argv, std::function<void(Server::Instance&)> post_server_create) {
  ... What we do now
}

int main(int argc, char** argv) {
  return main_helper(argc, argv, nullptr);
}

int hot_restart_main(int argc, char** argv) {
  return main_helper(argc, argv, [](Server::Instance&) { ... });
}

?
```

Reply:
I'm OK with that but I feel like this makes there be too many levels of "main" and I'd want to put this into MainCommon.

I feel like that will grow into a larger refactor and wonder if you'd be OK with the duplicated code for now and a TODO to refactor this into MainCommon?

Reply:
Sure that's fine to take a TODO or it doesn't seem too hard to just do it in MainCommon now but up to you.

Reply:
I've added a TODO and opened a thread with the absl communithy about the implications of initializing the symbolizer indirectly from main().